watcher/src/watcher_db.rs

```diff
@@ -0,0 +1,252 @@
+    counts: Database,
+
+    /// Logger.
+    logger: Logger,
+}
+
+impl WatcherDB {
+    /// Open an existing WatcherDB.
+    pub fn open(path: PathBuf, logger: Logger) -> Result<Self, WatcherDBError> {
+        let env = Arc::new(
+            Environment::new()
+                .set_max_dbs(10)
+                .set_map_size(MAX_LMDB_FILE_SIZE)
+                .open(path.as_ref())?,
+        );
+        let counts = env.open_db(Some(COUNTS_DB_NAME))?;
+        let block_signatures = env.open_db(Some(BLOCK_SIGNATURES_DB_NAME))?;
+
+        Ok(WatcherDB {
+            env,
+            block_signatures,
+            counts,
+            logger,
+        })
+    }
+
+    /// Create a fresh WatcherDB.
+    pub fn create(path: PathBuf) -> Result<(), WatcherDBError> {
+        let env = Arc::new(
+            Environment::new()
+                .set_max_dbs(10)
+                .set_map_size(MAX_LMDB_FILE_SIZE)
+                .open(path.as_ref())?,
+        );
+
+        let counts = env.create_db(Some(COUNTS_DB_NAME), DatabaseFlags::empty())?;
+        env.create_db(
+            Some(BLOCK_SIGNATURES_DB_NAME),
+            DatabaseFlags::DUP_SORT | DatabaseFlags::DUP_FIXED,
+        )?;
+
+        let mut db_txn = env.begin_rw_txn()?;
+        db_txn.put(counts, &NUM_BLOCKS_KEY, &encode(&0), WriteFlags::empty())?;
+        db_txn.commit()?;
+        Ok(())
+    }
+
+    /// Add a signature for a block.
+    pub fn add_block_signatures(
+        &self,
+        block_index: u64,
+        signatures: &[BlockSignature],
+    ) -> Result<(), WatcherDBError> {
+        // NOTE: num_blocks is actually "block height" - some blocks may be missing.
+        // This allows us to skip the origin block, and assumes we are adding signatures
+        // to monotonically increasing block indices.
+        let mut db_txn = self.env.begin_rw_txn()?;
+        let num_blocks = self.num_blocks()?;
+        if block_index >= num_blocks {
+            db_txn.put(
+                self.counts,
+                &NUM_BLOCKS_KEY,
+                &encode(&(block_index + 1)),
+                WriteFlags::empty(),
+            )?;
+        } else {
+            log::error!(
+                self.logger,
+                "Can not add new block at {:?}. Height is {:?}",
+                block_index,
+                num_blocks
+            );
+            return Err(WatcherDBError::BlockOrder);
+        }
+
+        let key_bytes = encode(&block_index);
+        for signature in signatures {
+            let value_bytes = encode(signature);
+            db_txn.put(
+                self.block_signatures,
+                &key_bytes,
+                &value_bytes,
+                WriteFlags::empty(),
+            )?;
+        }
+        db_txn.commit()?;
+        Ok(())
+    }
+
+    /// Get the signatures for a block.
+    pub fn get_block_signatures(
+        &self,
+        block_index: u64,
+    ) -> Result<Vec<BlockSignature>, WatcherDBError> {
```

Comment:
Commenting here since this is where the thought occurred to me - it might be nice to store the block URL the signature was obtained from? That could allow us to see how nodes change signatures over time. I'd wrap `BlockSignature` in a struct that contains it + the base URL + the filename.

Reply:
Yep - this is the plan for addressing the next phase of the block explorer, that tracks the change of the signature over time per enclave. I'm planning to do that as part of the next PR.

Reply:
Took care of the plumbing for this in the current PR - will use the data in the next PR.